badger_db.go

```diff
@@ -0,0 +1,292 @@
+		iter.Next()
+	}
+	return &badgerDBIterator{
+		reverse: opts.Reverse,
+		start:   start,
+		end:     end,
+
+		txn:  txn,
+		iter: iter,
+	}, nil
+}
+
+func (b *BadgerDB) Iterator(start, end []byte) (Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	return b.iteratorOpts(start, end, opts)
+}
+
+func (b *BadgerDB) ReverseIterator(start, end []byte) (Iterator, error) {
+	opts := badger.DefaultIteratorOptions
+	opts.Reverse = true
+	return b.iteratorOpts(end, start, opts)
+}
+
+func (b *BadgerDB) Stats() map[string]string {
+	return nil
+}
+
+func (b *BadgerDB) NewBatch() Batch {
+	wb := &badgerDBBatch{
+		db:         b.db,
+		wb:         b.db.NewWriteBatch(),
+		firstFlush: make(chan struct{}, 1),
+	}
+	wb.firstFlush <- struct{}{}
+	return wb
+}
+
+var _ Batch = (*badgerDBBatch)(nil)
+
+type badgerDBBatch struct {
+	db *badger.DB
+	wb *badger.WriteBatch
+
+	// Calling db.Flush twice panics, so we must keep track of whether we've
+	// flushed already on our own. If Write can receive from the firstFlush
+	// channel, then it's the first and only Flush call we should do.
+	//
+	// Upstream bug report:
+	// https://github.com/dgraph-io/badger/issues/1394
+	firstFlush chan struct{}
+}
+
+func (b *badgerDBBatch) Set(key, value []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	if value == nil {
+		return errValueNil
+	}
+	return b.wb.Set(key, value)
+}
+
+func (b *badgerDBBatch) Delete(key []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	return b.wb.Delete(key)
+}
+
+func (b *badgerDBBatch) Write() error {
+	select {
+	case <-b.firstFlush:
+		return b.wb.Flush()
+	default:
+		return fmt.Errorf("batch already flushed")
+	}
+}
+
+func (b *badgerDBBatch) WriteSync() error {
+	return withSync(b.db, b.Write())
+}
+
+func (b *badgerDBBatch) Close() error {
+	select {
+	case <-b.firstFlush: // a Flush after Cancel panics too
+	default:
+	}
+	b.wb.Cancel()
+	return nil
+}
+
+type badgerDBIterator struct {
+	reverse    bool
+	start, end []byte
+
+	txn  *badger.Txn
+	iter *badger.Iterator
+
+	lastErr error
+}
+
+func (i *badgerDBIterator) Close() error {
+	i.iter.Close()
+	i.txn.Discard()
+	return nil
+}
+
+func (i *badgerDBIterator) Domain() (start, end []byte) { return i.start, i.end }
+func (i *badgerDBIterator) Error() error                { return i.lastErr }
+
+func (i *badgerDBIterator) Next() {
+	if !i.Valid() {
+		panic("iterator is invalid")
+	}
+	i.iter.Next()
+}
+
+func (i *badgerDBIterator) Valid() bool {
+	if !i.iter.Valid() {
+		return false
+	}
+	if len(i.end) > 0 {
+		key := i.iter.Item().Key()
+		if c := bytes.Compare(key, i.end); (!i.reverse && c >= 0) || (i.reverse && c < 0) {
+			// We're at the end key, or past the end.
+			return false
+		}
+	}
+	return true
+}
+
+func (i *badgerDBIterator) Key() []byte {
+	if !i.Valid() {
+		panic("iterator is invalid")
+	}
+	// Note that we don't use KeyCopy, so this is only valid until the next
+	// call to Next.
+	return i.iter.Item().Key()
```

Comment:
We should use `KeyCopy()` here. The tm-db API only specifies that the key slice cannot be modified, so we shouldn't introduce backend-specific limitations like these as they can suddenly break applications when the user changes the backend.

Reply:
Thanks - I thought about it, but since the interface didn't specify that the `[]byte` are safe to use forever, I leaned towards the simpler (less expensive) option and the comment.

But what you say makes sense. I'll swap it. It's unfortunate that the interface has this performance limitation, though :)

Reply:
I agree, and scans are particularly performance-sensitive. We can consider adding this requirement, but it would require vetting a bunch of existing code. It would be useful with a realistic benchmark to back this change.

Reply:
Yeah, it sounds like a change that should be considered carefully at a later time. Definitely not for this PR - just thinking outloud.